File: ide-plugins/jetbrains/src/test/java/com/mercedesbenz/sechub/plugin/model/FileLocationExplorerTest.java
// SPDX-License-Identifier: MIT
package com.mercedesbenz.sechub.plugin.model;

import org.junit.Before;
import org.junit.Test;

import java.nio.file.Path;
import java.util.List;

import static com.mercedesbenz.sechub.plugin.model.TestResourceHelper.getEnsuredTestPath;
import static org.junit.Assert.*;

public class FileLocationExplorerTest {

    private FileLocationExplorer explorerToTest;

    @Before
    public void before() {
        explorerToTest = new FileLocationExplorer();
    }

    /* ----------------------------------------------- */
    /* ------------- REGEXP creation tests ----------- */
    /* ----------------------------------------------- */

    @Test
    public void path_contains_plusplus_is_correctly_handled() throws Exception{

        /* prepare */
        String location = "src/main/C++/os_injection.cpp";

        /* execute */
        String osSpecificLocation = explorerToTest.convertLocationOSSpecificRegExp(location, "\\");

        /* test */
        // remark: \Q start Quoting inside java regexp, \E ends it
        assertEquals("\\Qsrc\\main\\C++\\os_injection.cpp\\E",osSpecificLocation);
    }

    @Test
    public void location_converted_from_unix_to_windows_style_when_separator_is_backslash() throws Exception{

        /* prepare */
        String location = "src/main/java/com/example/Test1.java";

        /* execute */
        String osSpecificLocation = explorerToTest.convertLocationOSSpecificRegExp(location, "\\");

        /* test */
        assertEquals("\\Qsrc\\main\\java\\com\\example\\Test1.java\\E",osSpecificLocation);
    }

    @Test
    public void location_unconverted_from_unix_when_separator_is_slash() throws Exception{

        /* prepare */
        String location = "src/main/java/com/example/Test1.java";

        /* execute */
        String osSpecificLocation = explorerToTest.convertLocationOSSpecificRegExp(location, "/");

        /* test */
        assertEquals("\\Qsrc/main/java/com/example/Test1.java\\E",osSpecificLocation);
    }

    /* ----------------------------------------------- */
    /* ------------- SearchFor implementation tests -- */
    /* ----------------------------------------------- */
    @Test
    public void scenario1_projects_TestMe_java_found() throws Exception {
        /* prepare */
        Path project1 = getEnsuredTestPath("explorer/scenario1/project1");
        Path project2 = getEnsuredTestPath("explorer/scenario1/project2");
        Path expectedFile = getEnsuredTestPath("explorer/scenario1/project1/src/main/java/com/example/TestMe.java");

        explorerToTest.getSearchFolders().add(project1);
        explorerToTest.getSearchFolders().add(project2);

        /* execute */
        String locationString = "TestMe.java";
        List<Path> found = explorerToTest.searchFor(locationString);

        /* test */
        assertEquals(1,found.size());
        assertEquals(expectedFile, found.get(0));
    }

    @Test
    public void scenario1_projects_example_TestMe_java_found() throws Exception {
        /* prepare */
        Path project1 = getEnsuredTestPath("explorer/scenario1/project1");
        Path project2 = getEnsuredTestPath("explorer/scenario1/project2");
        Path expectedFile = getEnsuredTestPath("explorer/scenario1/project1/src/main/java/com/example/TestMe.java");

        explorerToTest.getSearchFolders().add(project1);
        explorerToTest.getSearchFolders().add(project2);

        /* execute */
        String locationString = "example/TestMe.java";
        List<Path> found = explorerToTest.searchFor(locationString);

        /* test */
        assertEquals(1,found.size());
        assertEquals(expectedFile, found.get(0));
    }

    @Test
    public void scenario1_projects_source_TestMe_c_found() throws Exception {
        /* prepare */
        Path project1 = getEnsuredTestPath("explorer/scenario1/project1");
        Path project2 = getEnsuredTestPath("explorer/scenario1/project2");
        Path expectedFile = getEnsuredTestPath("explorer/scenario1/project2/source/TestMe.c");

        explorerToTest.getSearchFolders().add(project1);
        explorerToTest.getSearchFolders().add(project2);

        /* execute */
        String locationString = "source/TestMe.c";
        List<Path> found = explorerToTest.searchFor(locationString);

        /* test */
        assertEquals(1,found.size());
        assertEquals(expectedFile, found.get(0));
    }

    @Test
    public void scenario1_projects_TestMe_c_found() throws Exception {
        /* prepare */
        Path project1 = getEnsuredTestPath("explorer/scenario1/project1");
        Path project2 = getEnsuredTestPath("explorer/scenario1/project2");
        Path expectedFile = getEnsuredTestPath("explorer/scenario1/project2/source/TestMe.c");

        explorerToTest.getSearchFolders().add(project1);
        explorerToTest.getSearchFolders().add(project2);

        /* execute */
        String locationString = "TestMe.c";
        List<Path> found = explorerToTest.searchFor(locationString);

        /* test */
        assertEquals(1,found.size());
        assertEquals(expectedFile, found.get(0));
    }

    @Test
    public void scenario1_projects_com_example_TestMe_java_found() throws Exception {
        /* prepare */
        Path project1 = getEnsuredTestPath("explorer/scenario1/project1");
        Path project2 = getEnsuredTestPath("explorer/scenario1/project2");
        Path expectedFile = getEnsuredTestPath("explorer/scenario1/project1/src/main/java/com/example/TestMe.java");

        explorerToTest.getSearchFolders().add(project1);
        explorerToTest.getSearchFolders().add(project2);

        /* execute */
        String locationString = "com/example/TestMe.java";
        List<Path> found = explorerToTest.searchFor(locationString);

        /* test */
        assertEquals(1,found.size());
        assertEquals(expectedFile, found.get(0));
    }

    @Test
    public void scenario1_projects_SameName_java_found() throws Exception {
        /* prepare */
        Path project1 = getEnsuredTestPath("explorer/scenario1/project1");
        Path project2 = getEnsuredTestPath("explorer/scenario1/project2");
        Path expectedFile1 = getEnsuredTestPath("explorer/scenario1/project1/src/main/java/com/example/SameName.java");
        Path expectedFile2 = getEnsuredTestPath("explorer/scenario1/project1/src/test/java/com/example/subpackage/SameName.java");
        Path unExpectedFile3 = getEnsuredTestPath("SameName.java");

        explorerToTest.getSearchFolders().add(project1);
        explorerToTest.getSearchFolders().add(project2);

        /* execute */
        String locationString = "SameName.java";
        List<Path> found = explorerToTest.searchFor(locationString);

        /* test */
        assertEquals(2,found.size());
        assertTrue(found.contains(expectedFile1));
        assertTrue(found.contains(expectedFile2));
        assertFalse(found.contains(unExpectedFile3));
    }

    @Test
    public void scenario2_projects_cpp_os_injection_cpp_found() throws Exception {
        /* prepare */
        Path project1 = getEnsuredTestPath("explorer/scenario2/project1");
        Path expectedFile = getEnsuredTestPath("explorer/scenario2/project1/Code/c++/os_injection.cpp");

        explorerToTest.getSearchFolders().add(project1);

        /* execute */
        String locationString = "os_injection.cpp";
        List<Path> found = explorerToTest.searchFor(locationString);

        /* test */
        assertEquals(1,found.size());
        assertEquals(expectedFile, found.get(0));
    }


}

File: ide-plugins/vscode/src/test/suite/filelocationexplorer.test.ts
// SPDX-License-Identifier: MIT
import * as assert from 'assert';
import * as path from 'path';
import * as fs from 'fs';

// You can import and use all API from the 'vscode' module
// as well as import your extension to test it
import { FileLocationExplorer } from '../../fileLocationExplorer';

suite('FileLocationExplorer Test Suite', () => {

	var explorerToTest: FileLocationExplorer;

	test('scenario1_projects_TestMe_java_found', () => {
		/* prepare */
		explorerToTest = new FileLocationExplorer();
		
		const project1 = getEnsuredTestPath("explorer/scenario1/project1");
		const project2 = getEnsuredTestPath("explorer/scenario1/project2");
		const expectedFilePath = getEnsuredTestPath("explorer/scenario1/project1/src/main/java/com/example/TestMe.java");

		explorerToTest.searchFolders.add(project1);
		explorerToTest.searchFolders.add(project2);

		/* execute */
		const locationString = "example/TestMe.java";
		var found = explorerToTest.searchFor(locationString);

		/* test */
		assert.strictEqual(1, found.size,"Not expected size of locations returned");
		const firstValue = found.values().next().value;
		assert.deepStrictEqual(firstValue,expectedFilePath,"must be expected result path");
	});

	test('scenario1_projects_source_TestMe_c_found', () => {
		/* prepare */
		explorerToTest = new FileLocationExplorer();
		
		const project1 = getEnsuredTestPath("explorer/scenario1/project1");
		const project2 = getEnsuredTestPath("explorer/scenario1/project2");
		const expectedFilePath = getEnsuredTestPath("explorer/scenario1/project2/source/TestMe.c");

		explorerToTest.searchFolders.add(project1);
		explorerToTest.searchFolders.add(project2);

		/* execute */
		const locationString = "source/TestMe.c";
		var found = explorerToTest.searchFor(locationString);

		/* test */
		assert.strictEqual(1, found.size,"Not expected size of locations returned");
		const firstValue = found.values().next().value;
		assert.deepStrictEqual(firstValue,expectedFilePath,"must be expected result path");

	});

	test('scenario1_projects_TestMe_c_found', () => {
		/* prepare */
		explorerToTest = new FileLocationExplorer();
		
		const project1 = getEnsuredTestPath("explorer/scenario1/project1");
		const project2 = getEnsuredTestPath("explorer/scenario1/project2");
		const expectedFilePath = getEnsuredTestPath("explorer/scenario1/project2/source/TestMe.c");

		explorerToTest.searchFolders.add(project1);
		explorerToTest.searchFolders.add(project2);

		/* execute */
		const locationString = "TestMe.c";
		var found = explorerToTest.searchFor(locationString);

		/* test */
		assert.strictEqual(1, found.size,"Not expected size of locations returned");
		const firstValue = found.values().next().value;
		assert.deepStrictEqual(firstValue,expectedFilePath,"must be expected result path");

	});

	test('scenario1_projects_com_example_TestMe_java_found', () => {
		/* prepare */
		explorerToTest = new FileLocationExplorer();
		
		const project1 = getEnsuredTestPath("explorer/scenario1/project1");
		const project2 = getEnsuredTestPath("explorer/scenario1/project2");
		const expectedFilePath = getEnsuredTestPath("explorer/scenario1/project1/src/main/java/com/example/TestMe.java");

		explorerToTest.searchFolders.add(project1);
		explorerToTest.searchFolders.add(project2);

		/* execute */
		const locationString = "com/example/TestMe.java";
		var found = explorerToTest.searchFor(locationString);

		/* test */
		assert.strictEqual(1, found.size,"Not expected size of locations returned");
		const firstValue = found.values().next().value;
		assert.deepStrictEqual(firstValue,expectedFilePath,"must be expected result path");

	});

	test('scenario1_projects_SameName_java_found', () => {
		/* prepare */
		explorerToTest = new FileLocationExplorer();
		
		const project1 = getEnsuredTestPath("explorer/scenario1/project1");
		const project2 = getEnsuredTestPath("explorer/scenario1/project2");
		const expectedFilePath1 = getEnsuredTestPath("explorer/scenario1/project1/src/main/java/com/example/SameName.java");
		const expectedFilePath2 = getEnsuredTestPath("explorer/scenario1/project1/src/test/java/com/example/subpackage/SameName.java");

		explorerToTest.searchFolders.add(project1);
		explorerToTest.searchFolders.add(project2);

		/* execute */
		const locationString = "SameName.java";
		var found = explorerToTest.searchFor(locationString);

		/* test */
		assert.strictEqual(2, found.size,"Not expected size of locations returned");
		const valuesIterator = found.values();
		const firstValue = valuesIterator.next().value;
		const secondValue = valuesIterator.next().value;
		assert.deepStrictEqual(firstValue,expectedFilePath1,"must be expected result path");
		assert.deepStrictEqual(secondValue,expectedFilePath2,"must be expected result path");

	});

	function getEnsuredTestPath(testfile: String): string {
		let testReportLocation = path.dirname(__filename) + "../../../../src/test/resources/" + testfile;
		var resolved = path.resolve(testReportLocation);
		if (!fs.existsSync(resolved)) {
			assert.fail("test report location does not exist:"+resolved);
		}
		return resolved;
	}

});
